src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+/// Provided a batching scheme, a service and batch settings
+/// this type will handle buffering events via the batching scheme
+/// and dispatching requests via the service based on either the size
+/// of the batch or a batch linger timeout.
+///
+/// # Acking
+///
+/// Service based acking will only ack events when all prior request
+/// batches have been acked. This means if sequential requests r1, r2,
+/// and r3 are dispatched and r2 and r3 complete, all events contained
+/// in all requests will not be acked until r1 has completed.
+pub struct BatchSink<S, B, Request, E = DefaultExecutor> {
+    service: ServiceSink<S, Request>,
+    batch: B,
+    settings: BatchSettings,
+    linger: Option<Delay>,
+    closing: bool,
+    exec: E,
+    _pd: PhantomData<Request>,
+}
+
+impl<S, B, Request> BatchSink<S, B, Request, DefaultExecutor>
+where
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    B: Batch<Output = Request>,
+{
+    pub fn new(service: S, batch: B, settings: BatchSettings, acker: Acker) -> Self {
+        BatchSink::with_executor(service, batch, settings, acker, DefaultExecutor::current())
+    }
+}
+
+impl<S, B, Request, E> BatchSink<S, B, Request, E>
+where
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    B: Batch<Output = Request>,
+    E: Executor,
+{
+    pub fn with_executor(
+        service: S,
+        batch: B,
+        settings: BatchSettings,
+        acker: Acker,
+        exec: E,
+    ) -> Self {
+        let service = ServiceSink::new(service, acker);
+
+        Self {
+            service,
+            batch,
+            settings,
+            linger: None,
+            closing: false,
+            exec,
+            _pd: PhantomData,
+        }
+    }
+
+    fn should_send(&mut self) -> bool {
+        self.closing || self.batch.len() >= self.settings.size || self.linger_elapsed()
+    }
+
+    fn linger_elapsed(&mut self) -> bool {
+        if let Some(delay) = &mut self.linger {
+            delay.poll().expect("timer error").is_ready()
+        } else {
+            false
+        }
+    }
+}
+
+impl<S, B, Request, E> Sink for BatchSink<S, B, Request, E>
+where
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    B: Batch<Output = Request>,
+    E: Executor,
+{
+    type SinkItem = B::Input;
+    type SinkError = crate::Error;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        if self.batch.len() >= self.settings.size {
+            trace!("batch full.");
+            self.poll_complete()?;
+
+            if self.batch.len() > self.settings.size {
+                debug!(message = "Buffer full; applying back pressure.", size = %self.settings.size, rate_limit_secs = 10);
```

Comment:
Should probably stick with the `batch` terminology here instead of `buffer`.

Reply:
:+1: